### ISSUE TITLE
`BroadcastGrad` is so time consuming

### ISSUE THREAD

**Issue by LyricZhao** [mindspore-ai/mindspore#65]
Hi, I'm currently working on a Molecular Dynamics simulation using MindSpore. My code is in https://github.com/LyricZhao/MindSpore-MD. I use plenty of small operators like `max` and `sqrt` to describe the simulation.

Later I use nvprof to profile the code and below are the results:
![ms_profiling](https://user-images.githubusercontent.com/35299455/89853181-533f8980-dbc3-11ea-8770-0256bcc5f6bb.png)

We can see that `BroadcastGrad` kernel takes 45.7% of the total time and then I find that the kernel is triggered by the backpropagation of `P.Maximum()` (https://github.com/LyricZhao/MindSpore-MD/blob/master/mindspore_impl.py#L64). However, it's really a small part of my application in speaking of the computation amount. Or to say, we can see that this operator is 33x slower than `F.sqrt`, which is unreasonable.

I refer to the implement of the kernel (https://github.com/mindspore-ai/mindspore/blob/master/mindspore/ccsrc/backend/kernel_compiler/gpu/cuda_impl/broadcast_grad_impl.cu#L45). In order to do the broadcast between 2 tensors of different sizes, the kernel function uses 3 divs, 15 muls and 4 mods to just calculate a max operation, which is not efficient.

In order to speed it up, I suggest that to rewrite the kernel or write different kernels for different cases (e.g. tensor and a scalar), or we can use TVM to generate the kernel after we know the exact size of the input.

p.s. Google has similar Molecular Dynamics simulation works (https://github.com/google/jax-md/) using JAX and TensorFlow-XLA compiler. Their solution is to just later compile the size and all the stuffs into a new fused kernel so that the complex loop could be optimized by backend compiler.

**Comment by WarmHouse** [comment]
Thank you for your issue. The scripts, profiling data and suggestions you provided are very helpful for problem location.

After analyzing the `MaximumGrad` operator, we found `AtomicAdd` the main source of time:
Since `Maximum` uses the matrix 'broadcast' in the forward operator, the gradient needs to be accumulated during the backward gradient propagation. 
In the your example, the `MaximumGrad` kernel has performed 4096*4096 AtomicAdd operations. The `AtomicAdd` is actually serial accumulation, which reduces the degree of parallelism.

Solution: 
On-demand gradient calculation is a fast and effective optimization solution for `MaximuimGrad` kernel. 
The results shows that the computer time is reduced from 26200us to 640us with 4096*4096. We will submit PR asap.
At the same time, graph kernel fusion(similar with XLA) will further improve the overall graph execution efficiency. It is under development.

**Comment by WilfChen** [comment]
Thank you for your issue. The scripts, profiling data and suggestions you provided are very helpful for problem location.

After analyzing the `MaximumGrad` operator, we found `AtomicAdd` the main source of time:
Since `Maximum` uses the matrix 'broadcast' in the forward operator, the gradient needs to be accumulated during the backward gradient propagation. 
In the your example, the `MaximumGrad` kernel has performed 4096*4096 AtomicAdd operations. The `AtomicAdd` is actually serial accumulation, which reduces the degree of parallelism.

Solution: 
On-demand gradient calculation is a fast and effective optimization solution for `MaximuimGrad` kernel. 
The results shows that the computer time is reduced from 26200us to 640us with 4096*4096 matrix. We will submit PR asap.
At the same time, graph kernel fusion(similar with XLA) will further improve the overall graph execution efficiency. It is under development.

**Comment by LyricZhao** [comment]
> Thank you for your issue. The scripts, profiling data and suggestions you provided are very helpful for problem location.
> 
> After analyzing the `MaximumGrad` operator, we found `AtomicAdd` the main source of time:
> Since `Maximum` uses the matrix 'broadcast' in the forward operator, the gradient needs to be accumulated during the backward gradient propagation.
> In the your example, the `MaximumGrad` kernel has performed 4096*4096 AtomicAdd operations. The `AtomicAdd` is actually serial accumulation, which reduces the degree of parallelism.
> 
> Solution:
> On-demand gradient calculation is a fast and effective optimization solution for `MaximuimGrad` kernel.
> The results shows that the computer time is reduced from 26200us to 640us with 4096*4096 matrix. We will submit PR asap.
> At the same time, graph kernel fusion(similar with XLA) will further improve the overall graph execution efficiency. It is under development.

Thanks for your nice reply! But I'm wondering what's a "on-demand gradient calculation"? Can you show me a simple code for demo?

**Comment by WilfChen** [comment]
For example:
```
y = P.Maximuim(x, 0)
```
Suppose `x` is weight or output of other operator, the gradient of `x` is faster cause it does not need `AtomicAdd`
Caclation gradient of `0` is time consuming, but fortunately the gradient of a constant is unnecesary.
